{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 4.1: splits vs. partitions\n",
    "\n",
    "A. What happens if you specify a delimiter that is not present in the **split** method? What about the **partition** method? Can you think of any reason(s) related to this difference that would why you might prefer to use one method versus another?\n",
    "\n",
    "B. What happens if you join a list using the empty string as a delimiter? Make a reasonable guess and then test this out.\n",
    "\n",
    "\n",
    "## Exercise 4.2: Split drills\n",
    "Take the string \"Molecular Paleontology Lab, Institute of Vertebrate Paleontology and Paleoanthropology, Chinese Academy of Sciences, Beijing, China\" and using something we learned, re-create the following lists.\n",
    "```python\n",
    "A. ['Molecular', 'Paleontology', 'Lab,', 'Institute', 'of', 'Vertebrate', 'Paleontology', 'and', 'Paleoanthropology,', 'Chinese', 'Academy', 'of', 'Sciences,', 'Beijing,', 'China']\n",
    "\n",
    "B. ['Molecular Paleontology Lab', ' Institute of Vertebrate Paleontology and Paleoanthropology', ' Chinese Academy of Sciences', ' Beijing', ' China']\n",
    "\n",
    "C.['Molecular Paleontology Lab, Institute', 'Vertebrate Paleontology and Paleoanthropology, Chinese Academy', 'Sciences, Beijing, China']\n",
    "\n",
    "D. ['Molecular Paleontology Lab, Institute of Vertebrate Paleontology and Paleoanthropology, Chinese Academy of Sciences, Beijing, China']\n",
    "\n",
    "E. \"China\"\n",
    "\n",
    "F. ['Molecular Paleontology Lab, Institute of Vertebrate Paleontology and ', 'anthropology, Chinese Academy of Sciences, Beijing, China']\n",
    "```\n",
    "\n",
    "## Exercise 4.2: Formatting and manipulating strings\n",
    "Take the string above and do the following:\n",
    "\n",
    "A. Make all letters upper case.\n",
    "\n",
    "B. Using the original string, switch the case of each letter. \n",
    "\n",
    "C. Using the original string, replace the word \"Institute of Vertebrate Paleontology and Paleoanthropology\" with \"IVPP\" and \"China\" with \"Zhongguo\", and \"Chinese Academy of Sciences\" with \"CAS\". \n",
    "\n",
    "D. Turn the string in part C into a list of each word/acronym, and then join every other word starting with the first into a string, separated by commas. \n",
    "\n",
    "E. Using the string in part D, replace all commas with a tab, \"\\t\".\n",
    "\n",
    "\n",
    "## Exercise 4.3: read()\n",
    "What does the method **read** do that is different from **readlines** and **readline**?? Use a combination of searching this up using \"?\", online searches, and test out a script using this method. \n",
    "\n",
    "\n",
    "## Exercise 4.4: Pairwise combos\n",
    "A. Within the four nucleotide bases \"AGCT\", there are six unique combinations. Write a script using a nested for loop that will write to a new file each of these pairs in a new line with a set number of dashes '------' separating the two.  (Ex. A------T)\n",
    "\n",
    "B. Then, reopen the file you made in part A, read in each of your pairs, and write them in reverse order, all in lower case. \n",
    "\n",
    "\n",
    "## Exercise 4.5: Baby's first Fasta parser, revisited\n",
    "Let's return to your script for reading in fasta files. You will update this script to make it more flexible. Currently, the script reads in three genes and three sequences, putting all such information into a dictionary. You will make two major changes:\n",
    "\n",
    "A. You will first adjust the inputting genes/sequences part of the script so that you can indefinitely keep adding genes and sequences. Make sure you specify a way to tell the program you no longer want to input these data and you want to move to the next part of the script. Remember, software often has to be used by others who didn't write the script! Make sure to add commands that make it clear what the user must do, without them reading the raw script!\n",
    "\n",
    "B. Upon finishing inputting data, you will write a new fasta file, containing the inputted data. Remember that a fasta file has the format:\n",
    "```\n",
    ">genename1\n",
    "AAGTTCT \n",
    ">genename2\n",
    "TTCACCATATAGGGGGGATATCCT\n",
    "```\n",
    "\n",
    "## Exercise 4.6: More Fasta file parsing\n",
    "Copy the text below into a text file in the resources/ folder and label it *seq.fasta*. \n",
    "```\n",
    ">gene1\n",
    "ATGAGACGTAGTGCCAGTAGCGCGATGTAGCG\n",
    "ATGACGCATGACGCGCGACGCGCGAGTGAGCC\n",
    "ATACGCACGCATTGGCA\n",
    ">gene2\n",
    "ATGTTCGACGCATACGACGCGCAGTACCAGCA\n",
    "ATGACGCACCGGGATACACGACGCGGATTTTT\n",
    "ACGCACCGAGATAGCATAAAAGACCATTAG\n",
    ">gene3\n",
    "TTATGGCACCCACTAGAGCCAGATTATTTTAAA\n",
    "```\n",
    "Write a script called *read_fasta.py* that will open this file, read the lines, and store the data as a dictionary keyed by gene with the sequences as values. Make sure the sequences are contiguous (i.e. contain no endline ('\\n') characters), and make sure to remove the > from the names of the genes.\n",
    "\n",
    "\n",
    "## Exercise 4.7: Neanderthals\n",
    "Here, the goal is to parse through two Neanderthal mtDNA sequences and count the number of differences between the two. The ones we will compare are [Mezmaiskaya 1](https://www.ncbi.nlm.nih.gov/nuccore/FM865411.1) and [El Sidron 1253](https://www.ncbi.nlm.nih.gov/nuccore/FM865409.1). These links lead to the NCBI database where you will retrieve the mtDNA sequences. At the top right, you will see the \"Send:\" button. Click it and then click \"File\" under \"Choose Destination\". Then click \"FASTA\" under \"Format\". Finally, click \"Create File\". \n",
    "\n",
    "Using UNIX commands, move these two files to the resources/ folder, and place them in the same fasta file (by concatenating) under a new name **neanderthals.fasta**. \n",
    "\n",
    "Write a script where you read this file, and parse through the sequence data, storing them in a list or dictionary (You could use code from your script in Exercise 4.6!). Then compare the two sequences to each other and determine the fraction of the mtDNA sequence where there are mutations (the two nucleotide bases differ). \n",
    "\n",
    "\n",
    "## Exercise 4.8: Getting real and old, Python style! (CHALLENGE!)\n",
    "Do you even remember Day 1 of lessons? In Exercise 1.7, parts C, D, and E. I asked you to use UNIX commands to find the number of each sex, the amount of missing data, and whether certain individuals had genotype calls or only allele calls. Can you write a python script that will do the equivalent of everything you did in UNIX?\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
