{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 1.1: Final installation check.\n",
    "If you have not gotten Python and the Jupyter Notebook to work (or your Terminal), then that is the first step! Please let your instructor know ASAP if this is an issue!\n",
    "\n",
    "## Exercise 1.2: Organizing and reviewing Unix\n",
    "A. If you haven't already, make a folder called PythonBootcamp/ in your home directory. Make sure there is an exercises/ folder, a lessons/ folder, a resources/ folder, and a Lesson1/ folder. Put the original set of lessons you download into the lessons/ folder, and original exercises into the exercises/ folder. Try using `mv`.  \n",
    "\n",
    "B. \n",
    "\n",
    "\n",
    "## Exercise 1.2: Cerevisiae chromosomes\n",
    "B. Go to [this link](http://downloads.yeastgenome.org/sequence/S288C_reference/chromosomes/fasta/) and download each of the files ending in .fsa. These are the chromosomes of the yeast, S. cerevisiae. You may have to right-click these files depending on your web browser (and be aware, some browsers will save your file with a .txt extension). If **wget** works in your Terminal, you can also use \n",
    "```bash\n",
    "wget http://downloads.yeastgenome.org/sequence/S288C_reference/chromosomes/fasta/*fsa\n",
    "```\n",
    "**curl** is another option to download direct from Terminal, where you can do:\n",
    "```bash\n",
    "curl http://downloads.yeastgenome.org/sequence/S288C_reference/chromosomes/fasta/chr01.fsa > chr01.fsa\n",
    "```\n",
    "to download all files directly into your resources/ folder.\n",
    "\n",
    "C. Make a single whole genome file called \"cerevisiae_genome.fasta\" by combining (concatenating) all the files into one file.\n",
    "\n",
    "D. Count the chromosomes in the whole genome file using commands from the lecture. (HINT 1: Each of the original FASTA files contains a single chromosome; HINT 2: Look at the additional options for **grep** - is there any that can count numbers of lines containing a search term? What about finding lines that do not contain a particular search term?).\n",
    "\n",
    "E. Look up the command 'wc' and find out what it does. Get size of total genome. (HINT: The size of the genome can be determined by counting the number of characters not on the same line as a fasta header).\n",
    "\n",
    "\n",
    "## Exercise 1.3: Cerevisiae genes\n",
    "A. Download a [list of cerevisiae chromosome features](http://downloads.yeastgenome.org/curation/chromosomal_feature/SGD_features.tab)\n",
    "\n",
    "**Columns within sgd_features.tab:**\n",
    "1. primary standfor gene database id (sgdid) (mandatory)\n",
    "2. feature type (mandatory)\n",
    "3. feature qualifier (optional)\n",
    "4. feature name (optional)\n",
    "5. standard gene name (optional)\n",
    "6. alias (optional, multiples separated by |)\n",
    "7. parent feature name (optional)\n",
    "8. secondary sgdid (optional, multiples separated by |)\n",
    "9. chromosome (optional)\n",
    "10. start_coordinate (optional)\n",
    "11. stop_coordinate (optional)\n",
    "12. strand (optional)\n",
    "13. genetic position (optional)\n",
    "14. coordinate version (optional)\n",
    "15. sequence version (optional)\n",
    "16. description (optional)\n",
    "\n",
    "B. Count the total number of genes (HINT 1: ORF indicates the entry is a gene).\n",
    "\n",
    "C. Count only verified genes. Count only uncharacterized genes.\n",
    "\n",
    "D. What other types of genes are in this file? For this, you may want to use the **sort -u** command, which will sort the input alphabetically, then take only unique lines.\n",
    "\n",
    "## Exercise 1.4: Star-struck\n",
    "From the same directory as your fasta files, see if you can predict what each of these commands will do (then try it):\n",
    "```bash\n",
    "head *\n",
    "head *.fsa\n",
    "head chr1*.fsa\n",
    "head chr1*\n",
    "head chr*1.fsa\n",
    "head chr*1\n",
    "grep 'S288C' *\n",
    "grep 'S288C' *.fsa\n",
    "grep 'BK006935.2' *\n",
    "cat * | grep 'BK006935.2' (what's the difference in the output between this one and the last one?)\n",
    "head *.fsa | grep 'chr'\n",
    "head *.fsa | grep 'chromosome' (what's the difference in the output between this one and the last one?)\n",
    "```\n",
    "\n",
    "## Exercise 1.5: Building a pipeline\n",
    "A. Download the two files [here](https://intro-prog-bioinfo-2015.wikispaces.com/file/view/palinsreg.txt/556091941/palinsreg.txt) and [here](https://intro-prog-bioinfo-2015.wikispaces.com/file/view/palinscmp.txt/556091935/palinscmp.txt). These are detected terminator sequences in the E. coli genome (using the program GeSTer, if you're curious).\n",
    "\n",
    "B. The command **grep '/G=[^ ]\\*' somefile** will find all lines that match */G=somegenename*, where somegenename is a sequence of non-blank characters. Read the output of **man grep** and figure out how to only print */G=somegenename*, rather than the whole line.\n",
    "\n",
    "C. Pipe the results of part B) through a **cut** command to get only everything after the =\n",
    "Store the results of part C) in a file named \"terminated_genes.txt'\n",
    "\n",
    "D. BONUS: Google for a Unix command that only keeps each gene once, rather than once per annotated terminator.\n",
    "\n",
    "## Exercise 1.6: Moving beyond the lecture\n",
    "A. Use Google and any other references you want to find a command that tells you how much disk space you have left.\n",
    "\n",
    "B. Use the **man** command to see how it works.\n",
    "\n",
    "C. How much space is left on your system? Find the option that makes the command output in terms of gigabytes and megabytes - the 'human-readable' form.\n",
    "\n",
    "## Exercise 1.7: Getting real and old (Challenge question - this may be difficult!)\n",
    "A. Let's look at a dataset a bit more relevant to our line of work. Go to [the Reich Lab website](https://reich.hms.harvard.edu/datasets), scroll down, and find the \"Genotypes of individuals\" dataset from the Fu et al. (2016) study. Make a directory titled fuetal2016/ in your resources/ directory and download the dataset into this folder. \n",
    "\n",
    "B. Use the command **unzip** to unzip the downloaded file. If you are using cygwin and it does not work, try using **apt-cyg install unzip** to download the Unix command. \n",
    "\n",
    "C. There are two sets of data, the **51.2.2M** and the **archaic**. These represent two different panels of SNPs that were captured for this study. Review the README for more information. The **.ind** file lists all the individuals, showing their ID, sex, and name. The **.snp** file lists the set of SNPs used (each row is one SNP, with information on the SNP ID, chromosome, genetic position, physical position, and the two alleles observed in humans. The **.geno** file shows each individual's genotype, with each row representing a SNP and each column representing an individual. The possible values are 0, 1, 2 and 9, where x = 0, 1 and 2 represent the number of the first allele observed in the individual (with 2-x being the number of the second allele observed in the individual) and 9 indicates missing data. \n",
    "\n",
    "Use a set of Unix commands to determine the number of males, females and individuals of unspecified sex in each dataset ( **51.2.2M** and **archaic**).\n",
    "\n",
    "D. All the included data are for ancient individuals, who may have missing data. What are the total number of SNPs used in each dataset? How many SNPs have no data for the UstIshim individual? For the Malta1 individual? For the Oase1 individual? For Vestonice16 or Loschbour?\n",
    "\n",
    "E. While this dataset is labeled as providing genotype information, is that actually true? A genotype implies you have knowledge of both alleles at each SNP (you know something about heterozygosity). Is that true here? How would you figure out if you do in fact have genotype data here, or if you only have single allele calls per SNP? Try this out for the individuals named in part D). \n",
    "\n",
    "## Exercise 1.8: Lazy...fractionator\n",
    "Let's move on to python! Set two variables **x** and **y** to any two numbers. Print out x/y in both fraction and decimal form.\n",
    "\n",
    "## Exercise 1.9: Temperature confusion\n",
    "A. Americans are a bit silly and use the Fahrenheit system, unlike most of the rest of the world. Thus, we often have trouble understanding the temperature when written in Celsius. Write me a script where you set a variable to a temperature in Celsius and it prints out the temperature in Fahrenheit. Then, write me a second script where it does the reverse. (Do this in a cell in the notebook)\n",
    "\n",
    "B. Write these two scripts to two files labeled *C2F.py* and *F2C.py*. Run these two programs through the Terminal.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
